Add metrics on the agent that can show connections status between agent with collections

WalkerWang731 opened the pull request:
<!--
We appreciate your contribution to the Jaeger project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- You have read the guide for contributing
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md#certificate-of-origin---sign-your-work
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Resolves #123"
-->

## Which problem is this PR solving?
- Resolves #2656 

## Short description of the changes
- Add `cmd/agent/app/reporter/connect_metrics.go`, used for add new metrics that connections status between agent with collection
- Add `cmd/agent/app/reporter/connect_metrics_test.go`, used for unit test to check func in `cmd/agent/app/reporter/connect_metrics.go`
- Adjust `cmd/agent/app/reporter/grpc/builder.go` func `CreateConnection()`, used for support change metrics as gRPC is changed
- Adjust `cmd/agent/app/reporter/grpc/collector_proxy.go` func `NewCollectorProxy()`, used for `CreateConnection()` newly added parameter requirements
- Adjust `cmd/agent/app/reporter/grpc/builder_test.go`, used for unit test to check func in `cmd/agent/app/reporter/grpc/builder.go` and newly added parameter requirements

## Changelog
Add metrics on the agent that can show connections status between the agent with collections

WalkerWang731 commented:
Below is my local test:

```
(cmd/agent/app)$ go test -v
=== RUN   TestAgentStartError
--- PASS: TestAgentStartError (0.00s)
=== RUN   TestAgentSamplingEndpoint
--- PASS: TestAgentSamplingEndpoint (0.51s)
=== RUN   TestAgentMetricsEndpoint
--- PASS: TestAgentMetricsEndpoint (0.51s)
=== RUN   TestStartStopRace
    agent_test.go:175: stopping agent
    logger.go:130: 2020-11-26T22:34:19.699+0800	INFO	shutting down agent's HTTP server	{"addr": "[::]:5778"}
    logger.go:130: 2020-11-26T22:34:19.699+0800	INFO	Starting jaeger-agent HTTP server	{"http-port": 5778}
    logger.go:130: 2020-11-26T22:34:19.700+0800	INFO	agent's http server exiting
--- PASS: TestStartStopRace (0.00s)
=== RUN   TestBuilderFromConfig
--- PASS: TestBuilderFromConfig (0.00s)
=== RUN   TestBuilderWithExtraReporter
--- PASS: TestBuilderWithExtraReporter (0.00s)
=== RUN   TestBuilderWithProcessorErrors
--- PASS: TestBuilderWithProcessorErrors (0.00s)
=== RUN   TestMultipleCollectorProxies
[{} {}]
--- PASS: TestMultipleCollectorProxies (0.00s)
=== RUN   TestCreateCollectorProxy
--- PASS: TestCreateCollectorProxy (0.21s)
=== RUN   TestCreateCollectorProxy_UnknownReporter
--- PASS: TestCreateCollectorProxy_UnknownReporter (0.00s)
=== RUN   TestPublishOpts
--- PASS: TestPublishOpts (0.00s)
=== RUN   TestBindFlags
--- PASS: TestBindFlags (0.00s)
PASS
ok  	github.com/jaegertracing/jaeger/cmd/agent/app	1.663s
```

```
(cmd/agent/app/reporter)$ go test -v
=== RUN   TestClientMetricsReporter_Zipkin
--- PASS: TestClientMetricsReporter_Zipkin (0.00s)
=== RUN   TestClientMetricsReporter_Jaeger
=== RUN   TestClientMetricsReporter_Jaeger/iter0
=== RUN   TestClientMetricsReporter_Jaeger/iter1
=== RUN   TestClientMetricsReporter_Jaeger/iter2
=== RUN   TestClientMetricsReporter_Jaeger/iter3
=== RUN   TestClientMetricsReporter_Jaeger/iter4
=== RUN   TestClientMetricsReporter_Jaeger/iter5
=== RUN   TestClientMetricsReporter_Jaeger/iter6
--- PASS: TestClientMetricsReporter_Jaeger (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter0 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter1 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter2 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter3 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter4 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter5 (0.00s)
    --- PASS: TestClientMetricsReporter_Jaeger/iter6 (0.00s)
=== RUN   TestClientMetricsReporter_ClientUUID
=== RUN   TestClientMetricsReporter_ClientUUID/iter0
=== RUN   TestClientMetricsReporter_ClientUUID/iter1
=== RUN   TestClientMetricsReporter_ClientUUID/iter2
=== RUN   TestClientMetricsReporter_ClientUUID/iter3
=== RUN   TestClientMetricsReporter_ClientUUID/iter4
=== RUN   TestClientMetricsReporter_ClientUUID/iter5
--- PASS: TestClientMetricsReporter_ClientUUID (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter0 (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter1 (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter2 (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter3 (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter4 (0.00s)
    --- PASS: TestClientMetricsReporter_ClientUUID/iter5 (0.00s)
=== RUN   TestClientMetricsReporter_Expire
=== RUN   TestClientMetricsReporter_Expire/detect_new_client
    client_metrics_test.go:261: gauge=1 detected on iteration 1
=== RUN   TestClientMetricsReporter_Expire/detect_stale_client
    client_metrics_test.go:278: gauge=0 detected on iteration 0
--- PASS: TestClientMetricsReporter_Expire (0.05s)
    --- PASS: TestClientMetricsReporter_Expire/detect_new_client (0.00s)
    --- PASS: TestClientMetricsReporter_Expire/detect_stale_client (0.05s)
=== RUN   TestConnectMetrics
--- PASS: TestConnectMetrics (0.00s)
=== RUN   TestBindFlags_NoJaegerTags
--- PASS: TestBindFlags_NoJaegerTags (0.00s)
=== RUN   TestBindFlags
--- PASS: TestBindFlags (0.00s)
=== RUN   TestBindFlagsAllInOne
--- PASS: TestBindFlagsAllInOne (0.00s)
=== RUN   TestMetricsReporter
--- PASS: TestMetricsReporter (0.00s)
=== RUN   TestMultiReporter
--- PASS: TestMultiReporter (0.00s)
=== RUN   TestMultiReporterErrors
--- PASS: TestMultiReporterErrors (0.00s)
PASS
ok  	github.com/jaegertracing/jaeger/cmd/agent/app/reporter	0.461s
```

```
(cmd/agent/app/reporter/grpc)$ go test -v
=== RUN   TestBuilderFromConfig
--- PASS: TestBuilderFromConfig (0.00s)
=== RUN   TestBuilderWithCollectors
=== RUN   TestBuilderWithCollectors/with_roundrobin_schema
=== RUN   TestBuilderWithCollectors/with_single_host
=== RUN   TestBuilderWithCollectors/with_custom_resolver_and_fixed_discoverer
=== RUN   TestBuilderWithCollectors/without_collectorPorts_and_resolver
=== RUN   TestBuilderWithCollectors/with_collector_connection_status_ready
=== RUN   TestBuilderWithCollectors/with_collector_connection_status_failure
--- PASS: TestBuilderWithCollectors (0.00s)
    --- PASS: TestBuilderWithCollectors/with_roundrobin_schema (0.00s)
    --- PASS: TestBuilderWithCollectors/with_single_host (0.00s)
    --- PASS: TestBuilderWithCollectors/with_custom_resolver_and_fixed_discoverer (0.00s)
    --- PASS: TestBuilderWithCollectors/without_collectorPorts_and_resolver (0.00s)
    --- PASS: TestBuilderWithCollectors/with_collector_connection_status_ready (0.00s)
    --- PASS: TestBuilderWithCollectors/with_collector_connection_status_failure (0.00s)
=== RUN   TestProxyBuilder
=== RUN   TestProxyBuilder/should_pass_with_insecure_grpc_connection
=== RUN   TestProxyBuilder/should_fail_with_secure_grpc_connection_and_a_CA_file_which_does_not_exist
=== RUN   TestProxyBuilder/should_pass_with_secure_grpc_connection_and_valid_TLS_Client_settings
--- PASS: TestProxyBuilder (0.00s)
    --- PASS: TestProxyBuilder/should_pass_with_insecure_grpc_connection (0.00s)
    --- PASS: TestProxyBuilder/should_fail_with_secure_grpc_connection_and_a_CA_file_which_does_not_exist (0.00s)
    --- PASS: TestProxyBuilder/should_pass_with_secure_grpc_connection_and_valid_TLS_Client_settings (0.00s)
=== RUN   TestProxyClientTLS
=== RUN   TestProxyClientTLS/should_pass_with_insecure_grpc_connection
=== RUN   TestProxyClientTLS/should_fail_with_TLS_client_to_non-TLS_server
=== RUN   TestProxyClientTLS/should_fail_with_TLS_client_to_untrusted_TLS_server
=== RUN   TestProxyClientTLS/should_fail_with_TLS_client_to_trusted_TLS_server_with_incorrect_hostname
=== RUN   TestProxyClientTLS/should_pass_with_TLS_client_to_trusted_TLS_server_with_correct_hostname
=== RUN   TestProxyClientTLS/should_fail_with_TLS_client_without_cert_to_trusted_TLS_server_requiring_cert
=== RUN   TestProxyClientTLS/should_fail_with_TLS_client_without_cert_to_trusted_TLS_server_requiring_cert_from_a_different_CA
=== RUN   TestProxyClientTLS/should_pass_with_TLS_client_with_cert_to_trusted_TLS_server_requiring_cert
--- PASS: TestProxyClientTLS (0.17s)
    --- PASS: TestProxyClientTLS/should_pass_with_insecure_grpc_connection (0.00s)
    --- PASS: TestProxyClientTLS/should_fail_with_TLS_client_to_non-TLS_server (0.14s)
    --- PASS: TestProxyClientTLS/should_fail_with_TLS_client_to_untrusted_TLS_server (0.00s)
    --- PASS: TestProxyClientTLS/should_fail_with_TLS_client_to_trusted_TLS_server_with_incorrect_hostname (0.00s)
    --- PASS: TestProxyClientTLS/should_pass_with_TLS_client_to_trusted_TLS_server_with_correct_hostname (0.00s)
    --- PASS: TestProxyClientTLS/should_fail_with_TLS_client_without_cert_to_trusted_TLS_server_requiring_cert (0.00s)
    --- PASS: TestProxyClientTLS/should_fail_with_TLS_client_without_cert_to_trusted_TLS_server_requiring_cert_from_a_different_CA (0.01s)
    --- PASS: TestProxyClientTLS/should_pass_with_TLS_client_with_cert_to_trusted_TLS_server_requiring_cert (0.01s)
=== RUN   TestMultipleCollectors
--- PASS: TestMultipleCollectors (0.00s)
=== RUN   TestBindFlags
--- PASS: TestBindFlags (0.00s)
=== RUN   TestReporter_EmitZipkinBatch
--- PASS: TestReporter_EmitZipkinBatch (0.00s)
=== RUN   TestReporter_EmitBatch
--- PASS: TestReporter_EmitBatch (0.00s)
=== RUN   TestReporter_SendFailure
--- PASS: TestReporter_SendFailure (0.00s)
=== RUN   TestReporter_AddProcessTags_EmptyTags
--- PASS: TestReporter_AddProcessTags_EmptyTags (0.00s)
=== RUN   TestReporter_AddProcessTags_ZipkinBatch
--- PASS: TestReporter_AddProcessTags_ZipkinBatch (0.00s)
=== RUN   TestReporter_AddProcessTags_JaegerBatch
--- PASS: TestReporter_AddProcessTags_JaegerBatch (0.00s)
=== RUN   TestReporter_MakeModelKeyValue
--- PASS: TestReporter_MakeModelKeyValue (0.00s)
PASS
ok  	github.com/jaegertracing/jaeger/cmd/agent/app/reporter/grpc	0.571s
```

objectiser commented:
@WalkerWang731 lint failed:

```
Go fmt, license check, or import ordering failures, run 'make fmt'
```

yurishkuro commented on cmd/agent/app/reporter/grpc/builder.go:
```diff
@@ -97,14 +100,27 @@ func (b *ConnBuilder) CreateConnection(logger *zap.Logger) (*grpc.ClientConn, er
 		logger.Info("Checking connection to collector")
 		for {
 			s := cc.GetState()
+			if s != connectivity.Ready {
```
nit: put `s == Ready` success case first.

WalkerWang731 replied:
For compatibility and unnecessary changes in the future. Actually, I concern the gRPC official changes connectivity status constant in the future version, although seems less possible.

yurishkuro commented on cmd/agent/app/reporter/grpc/builder_test.go:
```diff
@@ -149,7 +149,7 @@ func TestBuilderWithCollectors(t *testing.T) {
 			cfg.Discoverer = test.discoverer
 
-			conn, err := cfg.CreateConnection(zap.NewNop())
+			conn, err := cfg.CreateConnection(zap.NewNop(), metrics.NullFactory)
```
Please add a test for the new metrics. May need to refactor the builder a bit for better testability.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+	ExpireTTL       time.Duration
+}
+
+// ConnectMetricsReporter is a decorator also it not actual use currently.
```
this isn't accurate, it is not used as a decorator, and does not need to be called "reporter" since it's confusing in the current package where "reporter" refers to span reporter (or "exporter" in OTEL terminology).

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+	metric := r.params.MetricsFactory.Gauge(metrics.Options{
+		Name: "connected_collector_status",
+		Help: "Connection status that jaeger-agent to jaeger-collector, 1 is connected, 0 is disconnected",
+		Tags: map[string]string{"target": target},
```
The `target` here is the address of the collector, correct? Is it a good idea to use that as label, given that it could contain an ephemeral port number and prone to high cardinality? I'd say the target could be reported as an expvar variable, but the metric can be just the status.

jpkrohling replied:
> ephemeral port number

Care to expand on that? I don't remember seeing servers with ephemeral port numbers.

yurishkuro replied:
Why not? For example, at Uber most services run on ephemeral port numbers, to make it easy to stack and relocate them across available compute capacity. The discovery service reports host:port pairs for all instances of a given service. The custom build of the agent that Uber runs would use that discovery mechanism to locate Jaeger collectors. So the `target` string could be quite dynamic.

WalkerWang731 replied:
Actually, for this early version, I use such as `oldTarget` variable, the function incoming variable like `newTarget`, function will change `oldTarget` metric to 0, change `newTarget` metric to 1. In this case, we can watch which one agent connects to which collector. But eventually, I thought these are too complicated. We only need to observe the connection status is ok, after all. Which one agent connects to which collector should be added in collector metrics.

WalkerWang731 replied:
For `connected is 0 disconnected is 1` , because when have many connection metrics, we can filter `sum(jaeger_agent_connection_status_connected_collector_status{})  by (instance) > bool  0`  for quick view status.

I fixed the target, no need to pass the target parameter, in the next commit. target is static when the `connectMetrics{}` initialize

For the current PR, I thought that simplify the design first, and add new features if required.

jpkrohling replied:
> For example, at Uber most services run on ephemeral port numbers, to make it easy to stack and relocate them across available compute capacity

Didn't know that, but it does make sense for high density VMs/bare metal servers.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// CollectorConnected used for change metric as agent connected.
+func (r *ConnectMetricsReporter) CollectorConnected(target string) {
```
```suggestion
func (r *ConnectMetricsReporter) OnConnected(target string) {
```

yurishkuro replied:
Or perhaps have just a single func OnConnectionStatusChange and pass a parameter.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// CollectorAborted used for change metric as agent disconnected.
+func (r *ConnectMetricsReporter) CollectorAborted(target string) {
```
```suggestion
func (r *ConnectMetricsReporter) OnDisconnected(target string) {
```

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+	Logger          *zap.Logger     // required
+	MetricsFactory  metrics.Factory // required
+	ExpireFrequency time.Duration
+	ExpireTTL       time.Duration
```
It doesn't look like any of these fields are needed, let's simplify to only what's necessary. The gauge can be initialized in the constructor once (especially if `target` is not used as a label per my comment below).

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// WrapWithConnectMetrics creates ConnectMetricsReporter.
+func WrapWithConnectMetrics(params ConnectMetricsReporterParams) *ConnectMetricsReporter {
```
similar to comment above, this is not "wrapping" anything, just a standalone util class.

WalkerWang731 replied:
ok, I see, I also feel that there is no meaning, just maintain the same structure, let me adjust it to simplify.

yurishkuro reviewed (commented):
overall approach seems reasonable.

WalkerWang731 commented:
@yurishkuro @jpkrohling 

Hi all, 

I had committed code, about this commit I modified the following:

 - changed `cmd/agent/app/reporter/connect_metrics.go` relevant func that simplify useless parameters, and `target` is changed static when the `connectMetrics{}` initialize.
- changed `cmd/agent/app/reporter/connect_metrics.go` that add a new metric in `connectMetrics.ConnectedCollectorReconnect` used for can reflect current connection stability and demo the overall structure
 - changed `cmd/agent/app/reporter/grpc/builder.go` that `s == connectivity.Ready` 
-  changed `cmd/agent/app/reporter/grpc/builder.go` that add new variable `b.ConnectMetrics` used for unit test outside call
- changed `cmd/agent/app/reporter/grpc/builder_test.go` that add unit test in `TestBuilderWithCollectors()` as  builder unit test.
- changed `cmd/agent/app/reporter/connect_metrics_test.go` that add test new metric

yurishkuro commented on cmd/agent/app/reporter/grpc/builder_test.go:
```diff
@@ -149,7 +155,17 @@ func TestBuilderWithCollectors(t *testing.T) {
 			cfg.Discoverer = test.discoverer
 
-			conn, err := cfg.CreateConnection(zap.NewNop())
+			mb := metricstest.NewFactory(time.Hour)
```
`mf` is a more intuitive var name here

yurishkuro commented on cmd/agent/app/reporter/grpc/builder_test.go:
```diff
@@ -149,7 +155,17 @@ func TestBuilderWithCollectors(t *testing.T) {
+				MetricsFactory: mb,
+			}
+			cm := reporter.WrapWithConnectMetrics(params, test.target)
+			tr := &connectMetricsTest{
```
this struct serves no purpose, you can access the factory directly

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// WrapWithConnectMetrics creates ConnectMetrics.
+func WrapWithConnectMetrics(params ConnectMetricsParams, target string) *ConnectMetrics {
```
```suggestion
func NewConnectMetrics(params ConnectMetricsParams, target string) *ConnectMetrics {
```
nothing is being wrapped here

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	if params.ExpireTTL == 0 {
+		params.ExpireTTL = defaultExpireTTL
+	}
```
that is the point of ExpireFrequency/ExpireTTL?

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,82 @@
+// OnConnectionStatusChange used for pass the status parameter when connection is changed
+// 0 is disconnected, 1 is connected
+// For quick view status via use `sum(jaeger_agent_connection_status_connected_collector_status{}) by (instance) > bool 0`
+func (r *ConnectMetrics) OnConnectionStatusChange(status int64) {
```
this makes the API more explicit:
```suggestion
func (r *ConnectMetrics) OnConnectionStatusChange(connected bool) {
```

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,82 @@
+	if params.ExpireTTL == 0 {
+		params.ExpireTTL = defaultExpireTTL
+	}
+	params.Target = target
```
my objection was not about how target was passed, but to using it as a label in the metrics - it can cause high cardinality due to ephemeral ports.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,82 @@
+// 0 is disconnected, 1 is connected
+// For quick view status via use `sum(jaeger_agent_connection_status_connected_collector_status{}) by (instance) > bool 0`
+func (r *ConnectMetrics) OnConnectionStatusChange(status int64) {
+	metric := r.params.MetricsFactory.Gauge(metrics.Options{
```
This whole file uses just two metrics, but one initialized via reflection another explicitly here. Please pick just one way to simplify the code.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,70 @@
+// connectMetrics is real metric, but not support to directly change, need via ConnectMetrics for changed
+type connectMetrics struct {
+	// used for reflect current connection stability
+	ConnectedCollectorReconnect metrics.Counter `metric:"connected_collector_reconnect" help:"Default is 1, the metric can reflect current connection stability, as reconnect action increase the metric increase."`
```
```suggestion
	Reconnects metrics.Counter `metric:"collector_reconnects" help:"Number of successful connections (including reconnects) to the collector."`
```

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,70 @@
+	"go.uber.org/zap"
+)
+
+// connectMetrics is real metric, but not support to directly change, need via ConnectMetrics for changed
```
please remove this comment, it doesn't add much to the code

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,70 @@
+	ConnectedCollectorReconnect metrics.Counter `metric:"connected_collector_reconnect" help:"Default is 1, the metric can reflect current connection stability, as reconnect action increase the metric increase."`
+
+	// Connection status that jaeger-agent to jaeger-collector, 1 is connected, 0 is disconnected
+	ConnectedCollectorStatus metrics.Gauge `metric:"connected_collector_status" help:"Connection status that jaeger-agent to jaeger-collector, 1 is connected, 0 is disconnected"`
```
```suggestion
	Status metrics.Gauge `metric:"collector_connected" help:"Status of connection between the agent and the collector; 1 is connected, 0 is disconnected"`
```

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	cm := new(connectMetrics)
+	cmp.MetricsFactory = cmp.MetricsFactory.Namespace(metrics.NSOptions{Name: "connection_status"})
```
the extra namespacing is unnecessary here, metric names are already descriptive.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// NewConnectMetrics creates ConnectMetricsParams.
+func NewConnectMetrics(cmp ConnectMetricsParams) *ConnectMetricsParams {
```
the `cmp` struct is not needed, the only argument this function needs it the metrics factory, which can be passed directly.

the returned struct should not be called `...Params`, it's very confusing. Call it `ConnectMetrics`. It can be the same struct as `connectMetrics`.

WalkerWang731 replied:
ok let me adjust it

WalkerWang731 commented:
@jpkrohling @yurishkuro 

Hi Sorry for disturbing
I had committed code. 
According to @yurishkuro suggestion, about this commit I modified the following:
- changed `cmd/agent/app/reporter/connect_metrics.go` that integrated `ConnectMetrics()` and removed `ConnectMetricsParams()`
- changed `cmd/agent/app/reporter/grpc/builder.go` that achieve target by `expvar` and adjusted code of dependency relevant
- changed `cmd/agent/app/reporter/grpc/builder_test.go` and `cmd/agent/app/reporter/connect_metrics_test.go` ensure unit test is passed

btw: about the `Unit Tests / unit-tests (pull_request)` and `Build all-in-one / all-in-one (pull_request)`,  I saw doesn't pass. I don't know the reason. if need me to modify, let me know.

Thanks!

WalkerWang731 commented:
@yurishkuro @jpkrohling @Ashmita152
Hi all, who can take a look that why does build docker-images always failed recently?
I saw `scripts/travis/build-all-in-one-image.sh` was modified recently. I don't know whether related.

Ashmita152 commented:
Hi @WalkerWang731 Sorry for the trouble. You're right, we changed that build script recently and that break your build.

We missed an edge case where the PR is created from fork's master branch like yours this one. I have raised a fix: https://github.com/jaegertracing/jaeger/pull/2675 which should solve your build problem.

WalkerWang731 commented:
Hi @Ashmita152 
Thanks for the response. But now I encounter a new issue that doesn't pass `CIT Elasticsearch / elasticsearch 5.x (pull_request)`, I don't know who can help me take a look, and I also don't know whether related to recently modified.

If you are not responsible, but you know who can help me, can help me `@` 

thanks~~

Ashmita152 commented:
Hi @WalkerWang731 

The `CIT Elasticsearch / elasticsearch 5.x` job issue has been resolved in master. Thank you.

WalkerWang731 commented:
Hi @Ashmita152 really thanks for your response. 

But now, I encounter a new issue that doesn't pass `Unit Tests / unit-tests (pull_request) Failing after 7m — unit-tests`.  haha. 
Different from before, this case will sometimes fail sometimes succeed.
I encountered it in a previous test, but then passed as  `Merge branch 'master' into master`.


Above for your reference.

Same previous, if you are not responsible, but you know who can help me, can help me `@`


Thank you~

yurishkuro commented:
Must be another flaky test - #2679. I restarted the job.

WalkerWang731 commented:
> Must be another flaky test - #2679. I restarted the job.

ok, thanks, currently, looks all normal.
For coding, I had modified as required, if any questions let me know.

Thank you!

WalkerWang731 commented:
@jpkrohling Hi, Happy new year, may I know, how's this PR progress so far?

jpkrohling commented:
I see that there are a few comments from @yurishkuro that weren't addressed yet.

WalkerWang731 commented:
> I see that there are a few comments from @yurishkuro that weren't addressed yet.

Ah sorry, I thought it was pending for some reason.
Thanks for your reminder, I will fix it as soon as possible in few days.

yurishkuro commented on cmd/agent/app/reporter/connect_metrics.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// NewConnectMetrics will be initialize ConnectMetrics
+func (r *ConnectMetrics) NewConnectMetrics() {
```
This is unusual construct. I would make it simply a constructor function
```go
func NewConnectMetrics(logger *zap.Logger, metrics metrics.Factory) *ConnectMetrics
```

WalkerWang731 replied:
Ah, sorry for responding too late..

I saw that your PR https://github.com/jaegertracing/jaeger/pull/2728 and thanks for your help clean-up.

Looked forward to learning more from you

yurishkuro commented on cmd/agent/app/reporter/grpc/builder.go:
```diff
@@ -97,14 +104,38 @@ func (b *ConnBuilder) CreateConnection(logger *zap.Logger) (*grpc.ClientConn, er
+	go func(cc *grpc.ClientConn, cm *reporter.ConnectMetrics) {
 		logger.Info("Checking connection to collector")
+		var egt *expvar.String
+		r := expvar.Get("gRPCTarget")
+		if r == nil {
```
```suggestion
		if r := expvar.Get("gRPCTarget"); r == nil {
```

yurishkuro commented on cmd/agent/app/reporter/grpc/builder.go:
```diff
@@ -97,14 +104,38 @@ func (b *ConnBuilder) CreateConnection(logger *zap.Logger) (*grpc.ClientConn, er
+
+	go func(cc *grpc.ClientConn, cm *reporter.ConnectMetrics) {
 		logger.Info("Checking connection to collector")
+		var egt *expvar.String
```
why not move this into ConnectMetrics struct as well?

yurishkuro commented:
minor clean-up in #2728
